Refactor Build + Docker Uberjar Job

iethree opened the pull request:
[test run](https://github.com/metabase/metabase/actions/runs/14717704289?pr=57298)
[dockerhub upload](https://hub.docker.com/layers/metabase/metabase-dev/refactor-uberjar-job/images/sha256-c4bc00cd9852968618e72bb36ab15f42d48825b7e8b4c35927dbc00a5f56b4ec)
## Description

As this job has grown over time, it has become an unmaintainable mess

This job can be triggered in 2 ways
- pushing a (substantive) commit to one of the specified branches
- running a workflow dispatch action

This job does a few things
- Builds a jar
- Containerizes the jar
- Pushes the container to dockerhub
- Runs Trivy Vulnerability scanning

The prior version of the job had a bunch of branching logic to do all this

e.g.

```yaml
      - name: Run Trivy vulnerability scanner if master (ee)
        if: ${{ (github.ref_name == 'master') && matrix.edition == 'ee' }}
        uses: aquasecurity/trivy-action@0.29.0
        env:
          TRIVY_OFFLINE_SCAN: true
        with:
          image-ref: docker.io/${{ github.repository_owner }}/metabase-enterprise-head:latest
          format: sarif
          output: trivy-results.sarif
          version: "v0.57.1"

      - name: Run Trivy vulnerability scanner if master (oss)
        if: ${{ (github.ref_name == 'master') && matrix.edition == 'oss' }}
        uses: aquasecurity/trivy-action@0.29.0
        env:
          TRIVY_OFFLINE_SCAN: true
        with:
          image-ref: docker.io/${{ github.repository_owner }}/metabase-head:latest
          format: sarif
          output: trivy-results.sarif
          version: "v0.57.1"

      - name: Run Trivy vulnerability scanner if dev branch
        if: ${{ !(startsWith(github.ref_name,'master') || startsWith(github.ref_name,'backport')) && matrix.edition == 'ee' }}
        uses: aquasecurity/trivy-action@0.29.0
        env:
          TRIVY_OFFLINE_SCAN: true
        with:
          version: "v0.57.1"
          image-ref: docker.io/${{ github.repository_owner }}/metabase-dev:${{ steps.extract_branch.outputs.branch }}
          format: sarif
          output: trivy-results.sarif
```
The only difference between these three yaml blocks are the `image-ref` propery, which passes the container name

The main thing this refactor does is eliminate all of these duplicated blocks by simply generating docker repo and tag information once, and reusing them via variables. So instead of having 3 (or more) blocks doing the same thing, we just have one:

```yaml
      - name: Run Trivy vulnerability scanner
        uses: aquasecurity/trivy-action@0.29.0
        env:
          TRIVY_OFFLINE_SCAN: true
          repo: ${{ fromJson(needs.get-container-info.outputs[matrix.edition]).repo }}
          tag: ${{ fromJson(needs.get-container-info.outputs[matrix.edition]).tag }}
        with:
          image-ref: docker.io/${{ env.repo }}:${{ env.tag }}
          format: sarif
          output: trivy-results.sarif
          version: "v0.57.1"
```

I've also improved the efficiency of this job in a few ways
- removing some redundant checking of docker containers - these are just test builds, after all
- reusing an existing `containerize` workflow that builds multi-arch containers by default
- not building OSS jars for branches that don't push them to dockerhub

iethree committed:
rewrite uberjar workflow

iethree committed:
test

iethree committed:
fix

iethree committed:
fix

iethree committed:
fix

iethree committed:
fix

iethree committed:
fix

iethree committed:
fixes

iethree committed:
dont make docker sad

iethree committed:
remove test code

iethree commented on .github/workflows/containerize-jar.yml:
```diff
@@ -56,7 +56,7 @@ jobs:
           const version = "${{ inputs.tag }}".split('-')[0];
 
           if (!isValidVersionString(version)) {
-            return "linux/amd64";
+            return "linux/amd64,linux/arm64";
           }
```
this builds arm images by default for non-release builds

iethree committed:
remove metabot shenanigans

oreonl commented on .github/workflows/uberjar.yml:
```diff
@@ -23,16 +25,63 @@ on:
+    steps:
+      - name: Get docker repo and tag
+        id: info
+        uses: actions/script@v7
```
```suggestion
        uses: actions/github-script@v7
```

iethree replied:
wow. super weird [this actually worked](https://github.com/metabase/metabase/actions/runs/14673089003/job/41183912180?pr=57298). I guess I found a hidden alias? In any event, will use the right one.

oreonl reviewed (changes requested):
Other than the incorrect `uses` the only thing I'm not sure of is the Dockerhub credentials. The removed code used:
```
${{ secrets.DOCKERHUB_USERNAME }}
${{ secrets.DOCKERHUB_TOKEN }}
```
And the workflow being called now uses:
```
${{ secrets.DOCKERHUB_RELEASE_USERNAME }}
${{ secrets.DOCKERHUB_RELEASE_TOKEN }}
```
I'm not familiar with them and not sure they have the same permissions.

iethree committed:
address review comments

iethree committed:
different secrets handling

iethree committed:
remove test code

iethree commented:
> Other than the incorrect `uses` the only thing I'm not sure of is the Dockerhub credentials. The removed code used:
> 
> ```
> ${{ secrets.DOCKERHUB_USERNAME }}
> ${{ secrets.DOCKERHUB_TOKEN }}
> ```
> 
> And the workflow being called now uses:
> 
> ```
> ${{ secrets.DOCKERHUB_RELEASE_USERNAME }}
> ${{ secrets.DOCKERHUB_RELEASE_TOKEN }}
> ```
> 
> I'm not familiar with them and not sure they have the same permissions.

1. Fixed the actions script name
2. Updated the containerize jar so it can use either set of credentials, and will use the CI credentials for this workflow
